source/agora/common/BitMask.d

```diff
@@ -0,0 +1,230 @@
+/*******************************************************************************
+
+    This is an implementation of a bitmask that once initialized will have a
+    fixed number of bits for holding `true` / `false` (`1` / `0`) values.
+    It allocates the required number of `ubytes` in the constructor and does not
+    allow reading or writing to bits beyond the fixed count which is set during
+    construction.
+    This type is created for use as the validators signing bitmask and any
+    changes should ensure that it does not compromise that use.
+
+*******************************************************************************/
+
+module agora.common.BitMask;
+
+public struct BitMask
+{
+    import std.algorithm;
+    import std.conv;
+    import std.range;
+
+    @safe:
+
+    /// Pretty-print this value (to save bandwidth the shorter list is created)
+    /// `3/5 !(0,2)` can be read as 3 out of 5 with indices 0 and 2 not set (i.e. `01011`)
+    /// `2/5 (0,2)` can be read as 2 out of 5 with indices 0 and 2 set (i.e. `10100`)
+    /// If half are set we list the set bit indices e.g. `3/6 (0,1,2)` = `111000`
+    public void toString (scope void delegate (scope const char[]) @safe sink) const
+    {
+        import std.format;
+        bool list_set = this.setCount <= this.length / 2;
+        formattedWrite(sink, "%d/%d %s", this.setCount, this.length,
+            list_set
+                ? format!"(%s)"(this.setIndices.map!(i => i.to!string).join(","))
+                : format!"!(%s)"(this.notSetIndices.map!(i => i.to!string).join(",")));
+    }
+
+    /// Also support for Vibe.d serialization to JSON
+    public string toString () const
+    {
+        string ret;
+        scope void delegate (scope const char[]) @safe sink = (scope v) { ret ~= v; };
+        this.toString(sink);
+        return ret;
+    }
+
+    pure:
+
+    /// Support for Vibe.d deserialization
+    public static BitMask fromString (scope const(char)[] str)
+    {
+        auto set = str.findSplitBefore("/")[0].to!size_t;
+        auto length = str.findSplitAfter("/")[1].findSplitBefore(" ")[0].to!size_t;
+        bool list_set = set <= length / 2;
+        auto indices = str.findSplitAfter(list_set ? "(" : "!(")[1].findSplitBefore(")")[0]
+            .split(",").map!(i => i.to!size_t);
+        if (list_set)
+            assert(indices.length == set);
+        else
+            assert(set + indices.length == length);
+        auto bitmask = BitMask(length);
+        iota(length).each!((size_t i)
+        {
+            if (indices.canFind(i) == list_set)
+                bitmask[i] = true;
+        });
+        return bitmask;
```

Comment:
Ah this affects how Vibe.d serializes it right ?

Reply:
Now the `fromString` reads string of bit values (e.g. `0010101111`)